modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py

```diff
@@ -46,61 +47,72 @@ class PandasOnRayDataframeVirtualPartition(PandasDataframeAxisPartition):
+    def __init__(
+        self, list_of_partitions, get_ip=False, full_axis=True, call_queue=None
+    ):
+        if isinstance(list_of_partitions, PandasOnRayDataframePartition):
+            list_of_partitions = [list_of_partitions]
         self.full_axis = full_axis
         self.call_queue = call_queue or []
-        # In the simple case, none of the partitions that will compose this
-        # partition are themselves virtual partition. The partitions that will
-        # be combined are just the partitions as given to the constructor.
-        if not any(
-            isinstance(obj, PandasOnRayDataframeVirtualPartition)
-            for obj in list_of_blocks
-        ):
-            self.list_of_partitions_to_combine = list_of_blocks
-            return
-        # Check that all axis are the same in `list_of_blocks`
+        # Check that all virtual partition axes are the same in `list_of_partitions`
         # We should never have mismatching axis in the current implementation. We add this
         # defensive assertion to ensure that undefined behavior does not happen.
         assert (
             len(
                 set(
                     obj.axis
-                    for obj in list_of_blocks
+                    for obj in list_of_partitions
                     if isinstance(obj, PandasOnRayDataframeVirtualPartition)
                 )
             )
-            == 1
+            <= 1
         )
-        # When the axis of all virtual partitions matches this axis,
-        # extend and combine the lists of physical partitions.
-        if (
-            next(
-                obj
-                for obj in list_of_blocks
-                if isinstance(obj, PandasOnRayDataframeVirtualPartition)
-            ).axis
-            == self.axis
-        ):
-            new_list_of_blocks = []
-            for obj in list_of_blocks:
-                new_list_of_blocks.extend(
-                    obj.list_of_partitions_to_combine
-                ) if isinstance(
-                    obj, PandasOnRayDataframeVirtualPartition
-                ) else new_list_of_blocks.append(
-                    obj
-                )
-            self.list_of_partitions_to_combine = new_list_of_blocks
-        # Materialize partitions if the axis of this virtual does not match the virtual partitions
-        else:
-            self.list_of_partitions_to_combine = [
-                obj.force_materialization().list_of_partitions_to_combine[0]
-                if isinstance(obj, PandasOnRayDataframeVirtualPartition)
-                else obj
-                for obj in list_of_blocks
-            ]
+        self._list_of_constituent_partitions = list_of_partitions
+        # Defer computing _list_of_block_partitions because we might need to
+        # drain call queues for that.
+        self._list_of_block_partitions = None
+
+    @property
+    def list_of_block_partitions(self) -> List[partition_type]:
```

Comment:
this looks nearly identical to `list_of_block_partitions()` for `PandasOnDaskDataframeVirtualPartition`. Could it be combined in `PandasDataframeAxisPartition `?